src/NumericalAlgorithms/LinearOperators/IndefiniteIntegral.hpp

```diff
@@ -0,0 +1,148 @@
+ *
+ * \requires number of points in `integrand` and `mesh` are equal.
+ */
+template <size_t Dim>
+void indefinite_integral(gsl::not_null<DataVector*> integral,
+                         const DataVector& integrand,
+                         const Mesh<Dim>& mesh) noexcept;
+
+template <size_t Dim>
+DataVector indefinite_integral(const DataVector& integrand,
+                               const Mesh<Dim>& mesh) noexcept;
+// @}
+
+// @{
+/*!
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief Compute the indefinite integral of a function in the
+ * `dim_to_integrate`.
+ *
+ * The integral is done from the lowest dim to the highest dim, and after each
+ * integral the boundary condition is such that the integral vanishes at the
+ * logical lower end of the domain. For example, integrating \f$x^2y\f$
+ *
+ * One should be careful not to apply the boundary condition twice when using
+ * Gauss-Lobatto points. Consider the following element setup:
+ *
+ * \code
+ *  x - -x- - x - -x- - x
+ *  |         |         |
+ *  |    x    x    x    x
+ *  |2        |4        |
+ *  x - -x- - x - -x- - x
+ *  |         |         |   ^ y-axis
+ *  x    x    x    x    x   |
+ *  |1        |3        |   |
+ *  x - -x- - x - -x- - x   o----> x-axis
+ *  \endcode
+ *
+ * where `x` are the Gauss-Lobatto collocation points. The user would perform
+ * the integration in element `1` in the bottom left first, followed by elements
+ * `2` and `3` (upper left and bottom right, respectively), and finally in
+ * element `4` (top right). The boundary condition for the y-integral in element
+ * `2` is received from element `1` and care must be taken to not apply the
+ * x-integral more than once. That is, it could be applied once through the
+ * boundary condition and once via direct integration. In element `4` care must
+ * also be taken when applying boundary conditions.
+ *
+ * For Gauss points the grid structure is:
+ *
+ * \code
+ *  - - - - - - - - - - -
+ *  | x     x | x     x |
+ *  |    2    |    4    |
+ *  | x     x | x     x |
+ *  | - - - - - - -x- - |
+ *  | x     x | x     x |   ^ y-axis
+ *  |    1    |    3    |   |
+ *  | x     x | x     x |   |
+ *  - - - - - - - - - - -   o----> x-axis
+ *  \endcode
+ *
+ * where `x` are the Gauss-Lobatto collocation points. The integration proceeds
+ * with element `1` in the bottom left, followed by elements `2` and `3` (upper
+ * left and bottom right, respectively), and finally element `4` (top right).
+ *
+ * \requires number of points in `integrand` and `mesh` are equal.
+ */
+template <size_t Dim>
+void indefinite_integral(gsl::not_null<DataVector*> integral,
+                         const DataVector& integrand, const Mesh<Dim>& mesh,
+                         size_t dim_to_integrate) noexcept;
+
+template <size_t Dim>
+DataVector indefinite_integral(const DataVector& integrand,
+                               const Mesh<Dim>& mesh,
+                               size_t dim_to_integrate) noexcept;
+// @}
```

Comment:
More confused yet on boundary conditions, because unless I'm missing something badly, there don't appear to be any boundary conditions versions in the header or tests.